state-sync/state-sync-v2/state-sync-driver/src/metadata_storage.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    error::Error,
+    metadata_storage::database_schema::{MetadataKey, MetadataSchema, MetadataValue},
+};
+use anyhow::{anyhow, Result};
+use aptos_infallible::Mutex;
+use aptos_logger::prelude::*;
+use aptos_types::ledger_info::LedgerInfoWithSignatures;
+use schemadb::{
+    define_schema,
+    schema::{KeyCodec, ValueCodec},
+    ColumnFamilyName, Options, SchemaBatch, DB,
+};
+use serde::{Deserialize, Serialize};
+use std::{path::Path, sync::Arc, time::Instant};
+
+/// The metadata storage interface required by state sync to handle
+/// failures and reboots during critical parts of the syncing process.
+pub trait MetadataStorageInterface {
+    /// Returns true iff a state snapshot was successfully committed for the
+    /// specified target. If no snapshot progress is found, an error is returned.
+    fn is_snapshot_sync_complete(
+        &self,
+        target_ledger_info: &LedgerInfoWithSignatures,
+    ) -> Result<bool, Error>;
+
+    /// Gets the last persisted state value index for the snapshot sync at the
+    /// specified version. If no snapshot progress is found, an error is returned.
+    fn get_last_persisted_state_value_index(
+        &self,
+        target_ledger_info: &LedgerInfoWithSignatures,
+    ) -> Result<u64, Error>;
+
+    /// Returns the target ledger info of any state snapshot sync that has previously
+    /// started. If no snapshot sync started, None is returned.
+    fn previous_snapshot_sync_target(&self) -> Result<Option<LedgerInfoWithSignatures>, Error>;
+
+    /// Updates the last persisted state value index for the state snapshot
+    /// sync at the specified target ledger info.
+    fn update_last_persisted_state_value_index(
+        &self,
+        target_ledger_info: &LedgerInfoWithSignatures,
+        last_persisted_state_value_index: u64,
+        snapshot_sync_completed: bool,
+    ) -> Result<(), Error>;
+}
+
+/// The name of the state sync db file
+const STATE_SYNC_DB_NAME: &str = "state_sync_db";
+
+/// The name of the metadata column family
+const METADATA_CF_NAME: ColumnFamilyName = "metadata";
+
+/// A metadata storage implementation that uses a RocksDB backend to persist data
+#[derive(Clone)]
+pub struct PersistentMetadataStorage {
+    database: Arc<Mutex<DB>>,
```

Comment:
why this needs a Mutex? DB is thread-safe?

Reply:
Aah, thanks 😄 I wasn't initially sure how the `DB` would handle concurrency, so I played it safe 😄 I've removed the `Mutex`.